fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexKeyValueToPartialRecordUtils.java

```diff
@@ -0,0 +1,183 @@
+import com.apple.foundationdb.record.query.plan.IndexKeyValueToPartialRecord;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.foundationdb.tuple.TupleHelpers;
+import com.google.protobuf.Descriptors;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+
+/**
+ * A utility class to build a partial record for an auto-complete suggestion value, with grouping keys if there exist.
+ */
+public class LuceneIndexKeyValueToPartialRecordUtils {
+    private LuceneIndexKeyValueToPartialRecordUtils() {
+    }
+
+    public static void buildPartialRecord(@Nonnull KeyExpression root, @Nonnull Descriptors.Descriptor descriptor,
+                                          @Nonnull Message.Builder builder, @Nonnull String luceneField,
+                                          @Nonnull String suggestion) {
+        buildPartialRecord(root, descriptor, builder, luceneField, suggestion, TupleHelpers.EMPTY);
+    }
+
+    public static void buildPartialRecord(@Nonnull KeyExpression root, @Nonnull Descriptors.Descriptor descriptor,
+                                          @Nonnull Message.Builder builder, @Nonnull String luceneField,
+                                          @Nonnull String suggestion, @Nonnull Tuple groupingKey) {
+        final KeyExpression expression = root instanceof GroupingKeyExpression ? ((GroupingKeyExpression) root).getWholeKey() : root;
+        LuceneIndexExpressions.getFields(expression, new PartialRecordBuildSource(null, descriptor, builder),
+                (source, fieldName, value, type, stored, fieldNamePrefix, suffixOverride, groupingKeyIndex) -> {
+                    if (groupingKeyIndex > - 1) {
+                        if (groupingKeyIndex > groupingKey.size() - 1) {
+                            throw new RecordCoreException("Invalid grouping value tuple given a grouping key")
+                                    .addLogInfo(LogMessageKeys.VALUE, groupingKey.toString());
+                        }
+                        source.buildMessage(groupingKey.get(groupingKeyIndex), (String) value);
+                    } else if (type.equals(LuceneIndexExpressions.DocumentFieldType.TEXT) && fieldNameMatch(fieldName, luceneField, fieldNamePrefix, suffixOverride)) {
+                        source.buildMessage(suggestion, (String) value);
+                    }
+                    return;
+                },
+                null, 0, root instanceof GroupingKeyExpression ? ((GroupingKeyExpression) root).getGroupingCount() : 0);
+    }
+
+    private static boolean fieldNameMatch(@Nonnull String traversedFieldName, @Nonnull String givenFieldName,
+                                          @Nullable String fieldNamePrefix, boolean suffixOverride) {
+        // TODO: This is a workaround to deal with the issue that prefix of key path for grouping key could be missing, it will be fixed in a following commit
```

Comment:
I think it would be better to do an exact match on what is written today, rather than attempting a partial match.

I believe that means separating the grouping and grouped keys like `getGroupedFields` does before calling `LuceneIndexExpressions.getFields`. Or my making `LuceneIndexExpressions.getFields` take the root and have a more complicated contract that gives back the group as well.

This doesn't make it harder to change how the case of the group split in the middle of a nesting is handled later.

Reply:
But the ```LuceneIndexExpression.getFields``` just splits the whole key to grouping keys and grouped keys that the ```entry``` doesn't exist in the field name, for the test case of ```LuceneDocumentFromRecordTest.groupingMap()```. But if I understand correctly, we agree to fix it to be ```entry_value``` rather than ```value``` in the ```TODO``` followup work? @MMcM

Reply:
A separate PR to fix the existing issue got merge: https://github.com/FoundationDB/fdb-record-layer/pull/1514
Then this if block can be removed.